netket/nn/linear.py

```diff
@@ -207,6 +208,99 @@ def __call__(self, inputs: Array) -> Array:
         return y
 
 
+class DenseSymm(Module):
+    permutations: Any
+    alpha: Union[float, int]
+    use_bias: bool = True
+    dtype: Any = jnp.float64
+    precision: Any = None
+    kernel_init: Callable[[PRNGKey, Shape, Dtype], Array] = default_kernel_init
+    bias_init: Callable[[PRNGKey, Shape, Dtype], Array] = zeros
+
+    def setup(self):
+        perms = self.permutations()
+
+        self.n_symm, self.n_sites = perms.shape
+        self.features = int(self.alpha * self.n_sites / self.n_symm)
+        self.n_hidden = self.features * self.n_symm
+        if self.alpha > 0 and self.features == 0:
+            raise ValueError(
+                f"RBMSymm: alpha={self.alpha} is too small "
+                f"for {self.n_symm} permutations, alpha ≥ {self.n_symm / self.n_sites} is needed."
+            )
+
+        # symmetrization tensor, maps Wsymm to full W
+        # note that this should be a sparse object, once jax supports them
+        def symmetrizer_ijkl(i, j, k, l):
+            jsymm = np.floor_divide(j, self.n_symm)
+            cond_k = k == jnp.asarray(perms)[j % self.n_symm, i]
+            cond_l = l == jsymm
+            return jnp.asarray(jnp.logical_and(cond_k, cond_l), dtype=int)
+
+        self.symmetrizer = jnp.array(
+            np.fromfunction(
+                symmetrizer_ijkl,
+                shape=(self.n_sites, self.n_hidden, self.n_sites, self.features),
+                dtype=int,
+            )
+        )
+
+    def matmul(self, inputs: Array, kernel: Array) -> Array:
+        return lax.dot_general(
+            inputs,
+            kernel,
+            (((inputs.ndim - 1,), (0,)), ((), ())),
+            precision=self.precision,
+        )
+
+    @compact
+    def __call__(self, inputs: Array) -> Array:
+        """Applies a linear transformation to the inputs along the last dimension.
+        Args:
+          inputs: The nd-array to be transformed.
+        Returns:
+          The transformed input.
+        """
+        dtype = jnp.promote_types(inputs.dtype, self.dtype)
+        inputs = jnp.asarray(inputs, dtype)
+
+        kernel = self.param(
+            "kernel", self.kernel_init, (inputs.shape[-1], self.features), self.dtype
+        )
+        kernel = jnp.einsum("ijkl,kl->ij", self.symmetrizer, kernel)
```

Comment:
it might be worth benchmarking if doing jnp.vectorize-> matmul->reshape is more efficient than einsum.
It shouldn't be, but in the past I have seen that sometimes it is.

Reply:
I've compared `einsum` to this:
```python
kernel = jnp.matmul(self.symmetrizer, kernel.reshape(-1)).reshape(inputs.shape[-1], -1)
```
There does not seem to be much of a difference, at least for the Heisenberg1d example.
The `matmul` version might be a bit faster - but there is some variance so I'd need to look at it more carefully to make sure.